server/src/endpoint/writes.js

```diff
@@ -6,46 +6,135 @@ const r = require('rethinkdb');
+        response_data.forEach((data, index) => {
+          if (data === null) {
+            response_data[index] = new Error(timeout_msg);
+          }
+        });
+        return response_data;
+      }
+    }
+
+    return Promise.resolve().then(() => {
+      // Gather all rows to write
+      row_data.forEach((data) => {
+        if (data.version === undefined) {
+          delete data.row[hz_v];
         } else {
-          data[i] = { error: res.error };
-        }
-      } else if (res.new_val === null) {
-        data[i] = { id: res.old_val.id };
-        if (res.old_val[version_field] !== undefined) {
-          data[i][version_field] = res.old_val[version_field];
+          data.row[hz_v] = data.version;
         }
-      } else {
-        data[i] = { id: res.new_val.id };
-        data[i][version_field] = res.new_val[version_field];
+      });
+
+      if (ruleset.validation_required()) {
+        // For the set of rows to write, gather info for the validation step
+        return Promise.resolve(pre_validate(row_data.map((data) => data.row))).then((infos) => {
+          check(infos.length === row_data.length);
+
+          // For each row to write (and info), validate it with permissions
+          const valid_rows = [ ];
+          row_data.forEach((data, i) => {
+            const res = validate_row(data.row, infos[i]);
+
+            if (res !== undefined) {
+              response_data[data.index] = res;
+            } else {
+              valid_rows.push(data);
+            }
+          });
+          row_data = valid_rows;
+        });
       }
-    }
-  }
-  return { data, state: 'complete' };
+    }).then(() => { // For the set of valid rows, call the write step
+      if (row_data.length === 0) {
+        return [ ];
+      }
+      return do_write(row_data.map((data) => data.row)).then((res) => res.changes);
+    }).then((changes) => {
+      check(changes.length === row_data.length);
+
+      // Remove successful writes and invalidated writes that had an initial version
+      const retry_rows = [ ];
+      row_data.forEach((data, index) => {
+        const res = changes[index];
+        if (res.error !== undefined) {
+          if (res.error.indexOf('Duplicate primary key') === 0) {
+            response_data[data.index] = { error: 'The document already exists.' };
+          } else if (res.error.indexOf(invalidated_msg) === 0 &&
+                     data.version === undefined) {
+            retry_rows.push(data);
+          } else {
+            response_data[data.index] = { error: res.error };
+          }
+        } else if (res.new_val === null) {
+          response_data[data.index] = { id: res.old_val.id, [hz_v]: res.old_val[hz_v] };
+        } else {
+          response_data[data.index] = { id: res.new_val.id, [hz_v]: res.new_val[hz_v] };
+        }
+      });
+
+      row_data = retry_rows;
+
+      // Recurse, after which it will decide if there is more work to be done
+      return iterate();
```

Comment:
I moved `row_data`, `response_data`, and `timeout` down one scope to be parameters to the `iterate` call.  They are still global in the context of a single call to `iterate`, because chaining them from promise to promise is rather difficult, especially if one promise is from a ReQL query (as I'm sure you've encountered).  So, not much has changed, but I did what I could.